File: ai/common/lang.py
"""
Language detection utilities shared across AI service and HF Space.
Supports: Indonesian (id), Malay (ms), German (de), English (en default).
"""

from typing import Literal

Lang = Literal["id", "ms", "de", "en"]

ID_WORDS = [
    "kapan",
    "dimana",
    "di mana",
    "berapa",
    "harga",
    "acara",
    "bisnis",
    "saya",
    "tertarik",
    "manfaat",
    "jadwal",
]

MS_WORDS = [
    "bila",
    "di mana",
    "dimana",
    "berapa",
    "harga",
    "acara",
    "perniagaan",
    "syarikat",
    "nak",
    "penaja",
    "manfaat",
    "jadual",
    "hubungi",
]

DE_WORDS = ["wann", "wo", "wie", "veranstaltung", "preis", "geschäft", "hilfe"]


def detect_language(text: str) -> Lang:
    t = (text or "").lower()
    # token-based scoring to avoid false positives (e.g., English "sponsorship")
    import re
    tokens = set(re.findall(r"[a-z]+", t))

    def score(words):
        s = 0
        for w in words:
            if " " in w:
                if w in t:
                    s += 2  # phrase match gets higher weight
            else:
                if w in tokens:
                    s += 1
        return s

    id_s = score(ID_WORDS)
    ms_s = score(MS_WORDS)
    de_s = score(DE_WORDS)
    if id_s == 0 and ms_s == 0 and de_s == 0:
        return "en"
    # prefer the higher score; tie-breaking prefers id over ms over de
    if id_s >= ms_s and id_s >= de_s:
        return "id"
    if ms_s >= id_s and ms_s >= de_s:
        return "ms"
    return "de"


def get_language_prompt(language: Lang) -> str:
    if language == "id":
        return "Harap balas dalam Bahasa Indonesia (Bahasa Indonesia)."
    if language == "ms":
        return "Sila balas dalam Bahasa Melayu."
    if language == "de":
        return "Bitte auf Deutsch antworten."
    return "Please respond in English."

File: src/lib/ai/lang.ts
// Simple heuristic language detection tuned for ID/MS/EN/DE

export type Lang = "id" | "ms" | "en" | "de";

const ID_WORDS = [
  "kapan",
  "dimana",
  "di mana",
  "berapa",
  "harga",
  "acara",
  "bisnis",
  "saya",
  "tertarik",
  "penyelanggara",
  "manfaat",
  "jadwal",
];

const MS_WORDS = [
  "bila",
  "di mana",
  "berapa",
  "harga",
  "acara",
  "perniagaan",
  "syarikat",
  "saya",
  "nak",
  "penaja",
  "manfaat",
  "jadual",
  "hubungi",
];

const DE_WORDS = ["wann", "wo", "wie", "veranstaltung", "preis", "geschäft"];

export function detectLanguage(text: string): Lang {
  const t = (text || "").toLowerCase();
  const tokens = new Set((t.match(/[a-z]+/g) || []).filter(Boolean));
  const score = (words: string[]) =>
    words.reduce((s, w) => {
      if (w.includes(" ")) return s + (t.includes(w) ? 2 : 0); // phrase match gets higher weight
      return s + (tokens.has(w) ? 1 : 0);
    }, 0);

  const idScore = score(ID_WORDS);
  const msScore = score(MS_WORDS);
  const deScore = score(DE_WORDS);

  if (idScore === 0 && msScore === 0 && deScore === 0) return "en";
  if (idScore >= msScore && idScore >= deScore) return "id";
  if (msScore >= idScore && msScore >= deScore) return "ms";
  return "de";
}
